test/tesla/middleware/retry_test.exs

```diff
@@ -22,7 +23,32 @@ defmodule Tesla.Middleware.RetryTest do
   defmodule Client do
     use Tesla
 
-    plug Tesla.Middleware.Retry, delay: 10, max_retries: 10
+    plug Tesla.Middleware.Retry,
+      delay: 10,
+      max_retries: 10
+
+    adapter LaggyAdapter
+  end
+
+  defmodule ClientWithShouldRetryFunction do
+    use Tesla
+
+    plug Tesla.Middleware.Retry,
+      delay: 10,
+      max_retries: 10,
+      should_retry: fn
+        {:ok, %{status: status}} when status in [400, 500] ->
+          assert true
+          true
+
+        {:ok, _} ->
+          assert false
+          false
+
+        {:error, _} ->
+          assert false
+          true
+      end
```

Comment:
Are you sure this test works? Since you return `ok, 500` from the adapter the default `should_retry` implementation would not perform the retry and just return the `ok, 500`. 

Please take a look at the `"/maybe"` implementation - the adapter return value changes with the number of retries. There could be another `"/maybe-should-retry"` that would act in a similar way, but return 500 instead of `{:error, ...}`.